### ISSUE TITLE
File Attachments

### ISSUE THREAD

**Issue by gotdan** [HL7/bulk-data#63]
### Use case:
Handling of binary resources is not directly addressed in the IG leading to a lack of clarity on how servers and clients should handle NDJSON files with resources that contain attachments (eg. DocumentReferences)

### Proposed language:
If resources in an output file contain elements of the type ```Attachment```, servers SHALL populate the ```Attachment.contentType``` code as well as either the ```data``` element or the ```url``` element. The ```url``` element SHALL be either an absolute url that can be de-referenced to the attachment's content or a relative url that identifies a [Binary FHIR Resource](http://hl7.org/fhir/binary.html) included in the output of the export [*see Q1 below].

When the ```url``` element is populated with an absolute URL and the ```requiresAccessToken``` field in the Complete Status body is set to ```true```, the url location must be accessible by a client with a valid access token, and SHALL NOT require the use of additional authentication credentials.  When the ```url``` element is populated and the ```requiresAccessToken``` field in the Complete Status body is set to ```false```, the url location must be accessible by a client without an access token. 

Note that if a server copies files to the bulk data output endpoint or proxies requests to facilitate access from this endpoint, it may need to modify the ```Attachment.url``` element when generating the FHIR bulk data output files.

### Open Questions:
1.  Do we want to allow servers to represent attachments as relative references to Binary resources included in the export, or should we limit attachments to inline data and de-referenceable absolute URLs?
2. Server developers: Is the proposed language reasonably implementable?
3. Client developers: Does the proposed language address your needs with regards to file retrieval?

**Comment by vlad-ignatov** [comment]
`Binary` resources might be unreliable to consume as NDJSON if they contain big blobs of base64 `data`. On the client side, a streaming NDJSON parser needs to consume chunks of bytes and put them into internal buffer until an EOL (end of line) is reached. Only then can that line be parsed as JSON. If such parser is careful enough, it would have a limit for its buffer size. In other words, we can't handle a line that takes more memory than the amount available to the parser. This means that even is a server somehow manages to serialize all `Binary` resources as NDJSON, there is no guarantee that the clients can parse them.

**Comment by gotdan** [comment]
@vlad-ignatov - good point. I suppose one option would be to limit Binary resource in NDJSON to 1mb each or something. Curious if anyone sees value to this or if we should just require URL retrieval for all Binary resources, which would be my preference.

**Comment by jmandel** [comment]
I think that having a way to put all the data into the "Export pile" (i.e.,  throw this pile across the transom and be done) is important. But as Dan points out above, this can be accomplished by servers choosing to rewrite `Attachment.url` at export time (to point to a file in the exported pile, rather than a file "back" in the source of truth). Only details to think through on this front would be to make sure our advice on export time-to-live and cleanup are aligned and apply to any attachment files, just like to the Resource-based NDJSON files.

**Comment by prb112** [comment]
As the references/urls mutate, there would be a need (much like [$import](https://github.com/smart-on-fhir/bulk-import/blob/master/import.md#preserving-resource-ids) to have an extension to preserve provenance and source information:

```
"extension": [{
	"url": "http://hl7.org/fhir/StructureDefinition/export-source",
	"valueIdentifier": [{
		"system": "[exportSource]",
		"uri": "[original resource url"
	}]
}]
```

The same change could be made for identifier related to a reference.  
I also would add this as optional.

Thank you

**Comment by gotdan** [comment]
@prb112 Can you expand a bit on the use case? Are you talking about having the server record a history of changes in the download url for an attachment?  How would this be used by the client receiving the export? 

Also, not sure I follow your last point around multiple exports - can you clarify?

**Comment by dennispatterson** [comment]
Vlad's point is valid regarding parsing challenges introduced by embedding data as base64 in an ndjson file.  I don't have any issue (from the server perspective) with the approach of re-writing attachment urls to an exported url location.  I think this could be documented as the preferred approach, but not sure we need to outlaw the embedding approach if there are legitimate use-cases for smaller binaries...?  I'd be interested in hearing from client developers around their existing parser buffer sizes to help illustrate Vlad's concerns.

**Comment by prb112** [comment]
> @prb112 Can you expand a bit on the use case? Are you talking about having the server record a history of changes in the download url for an attachment? How would this be used by the client receiving the export?

Sure.  it's around provenance of `de-referenceable absolute URLs`.  once a resource is mutated, it is worthwhile to show the input source or originating URL.   The client is able to re-associate or trace a potential dangling reference.

> Also, not sure I follow your last point around multiple exports - can you clarify?

If you are doing an export from one system to another system and then another, it's worth managing a consistent reference.

**Comment by vlad-ignatov** [comment]
Just to clarify - while I believe my concern is valid, it is just a rare edge case that we need to consider. The default approach should work most of the time and we should not outlaw it.

If you have a FHIR server that can successfully serialize every single Binary resource, then it would be OK to export them all in on NDJSON. The total size of the NDJSON output doesn't really matter. It is the length of any single line that might be problematic.

That said, I think we need to keep the default approach and I agree that the server should be smart enough to rewrite it's output on the fly if it detects a "file" that is over a certain limit. This thing is just an improvement that we should settle on and document. A client may even decide not to support that (if it is only created for specific purpose and if it somehow knows that the server that it connects to does not have huge files).

**Comment by gotdan** [comment]
Following up on last week's workgroup meeting, I created a pull request updating the IG with the straight forward approach we discussed for the common case of exporting patient documents as well as an approach for returning "orphan" Binary resources. 

I didn't define extensions for retaining historical urls or for embedding metadata from a Binary Resource into an Attachment since I don't feel clear enough on the use cases involved to make a proposal. @prb112 - if you feel strongly that this should be in the IG, maybe draft a pull request that we can review at one of the workgroup meetings? Alternatively, we can let folks experiment with this as custom extensions in implementations and then go back and standardize whatever seems to be working best in the real world.
